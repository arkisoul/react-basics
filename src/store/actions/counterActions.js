export const INCREMENT = "INCREMENT";
export const DECREMENT = "DECREMENT";
export const INCREMENTBYVALUE = "INCREMENTBYVALUE";

// Action creators are functions that returns an action object
export const incrementCounter = () => ({ type: INCREMENT });
export const decrementCounter = () => ({ type: DECREMENT });
export const incrementCounterByValue = (value) => ({
  type: INCREMENTBYVALUE,
  payload: { value },
});
