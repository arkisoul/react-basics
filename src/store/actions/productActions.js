import axios from 'axios';

export const ADDPRODUCT = "ADDPRODUCT";
export const REMOVEPRODUCT = "REMOVEPRODUCT";
export const PRODUCT_REQUEST_START = "PRODUCT_REQUEST_START";
export const PRODUCT_REQUEST_SUCCESS = "PRODUCT_REQUEST_SUCCESS";
export const PRODUCT_REQUEST_FAILURE = "PRODUCT_REQUEST_FAILURE";

export const addProduct = (product) => ({
  type: ADDPRODUCT,
  payload: { product: product },
});
export const removeProduct = (productId) => ({
  type: REMOVEPRODUCT,
  payload: { productId: productId },
});
export const productFetchStart = () => ({ type: PRODUCT_REQUEST_START });
export const productFetchSuccess = (products) => ({
  type: PRODUCT_REQUEST_SUCCESS,
  payload: { products },
});
export const productFetchFailure = (error) => ({
  type: PRODUCT_REQUEST_FAILURE,
  payload: { error },
});

export const fetchProducts = () => async (dispatch) => {
  try {
    dispatch(productFetchStart());
    const res = await axios.get("https://fakestoreapi.com/products");
    const products = res.data;
    dispatch(productFetchSuccess(products))
  } catch (error) {
    dispatch(productFetchFailure(error))
  }
};
