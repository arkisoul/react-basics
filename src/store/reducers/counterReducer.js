import {
  INCREMENT,
  INCREMENTBYVALUE,
  DECREMENT,
} from "../actions/counterActions";

const counterInitialState = {
  count: 0,
};

export const counterReducer = (state = counterInitialState, {type, payload}) => {
  switch (type) {
    case INCREMENT:
      return { ...state, count: state.count + 1 };

    case DECREMENT:
      return { ...state, count: state.count - 1 };

    case INCREMENTBYVALUE:
      return { ...state, count: state.count + payload.value };

    default:
      return state;
  }
};