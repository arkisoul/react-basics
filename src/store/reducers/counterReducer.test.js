import { describe, it, expect } from "@jest/globals";

import { counterInitialState, counterReducer } from "./counterReducer";
import * as counterActions from "../actions/counterActions";

describe("Counter Reducer", () => {
  it("should return initial state on store creation", () => {
    expect(counterReducer(undefined, {})).toEqual(counterInitialState);
    expect(counterReducer(undefined, {})).toMatchSnapshot();
  });

  it("should increment the count value by 1 on INCREMENT action", () => {
    expect(
      counterReducer(undefined, { type: counterActions.INCREMENT })
    ).toEqual({
      count: 1,
    });
    expect(
      counterReducer({ count: 1 }, counterActions.incrementCounter())
    ).toEqual({
      count: 2,
    });
  });

  it("should decrement the count value by 1 on DECREMENT action", () => {
    expect(
      counterReducer(undefined, { type: counterActions.DECREMENT })
    ).toEqual({
      count: -1,
    });
    expect(
      counterReducer({ count: 4 }, counterActions.decrementCounter())
    ).toEqual({
      count: 3,
    });
  });

  it("should increment the count value by 10 on INCREMENTBYVALUE action", () => {
    const value = 10;
    expect(
      counterReducer(undefined, {
        type: counterActions.INCREMENTBYVALUE,
        payload: { value },
      })
    ).toEqual({
      count: 10,
    });
    expect(
      counterReducer(
        { count: 10 },
        counterActions.incrementCounterByValue(value)
      )
    ).toEqual({ count: 20 });
  });
});
