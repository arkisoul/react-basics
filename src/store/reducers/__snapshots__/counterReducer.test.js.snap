// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Reducer should return initial state on store creation 1`] = `
Object {
  "count": 0,
}
`;
